modules/apps/message-boards/message-boards-service/src/main/java/com/liferay/message/boards/internal/util/MBMessageNotificationTemplateHelper.java

```diff
@@ -0,0 +1,351 @@
+
+	public MBMessageNotificationTemplateHelper(
+		boolean htmlFormat, int maxNumberOfMessages,
+		int maxNumberOfThreadMessages,
+		MBMessageLocalService mbMessageLocalService,
+		ServiceContext serviceContext) {
+
+		_htmlFormat = htmlFormat;
+		_maxNumberOfMessages = maxNumberOfMessages;
+		_maxNumberOfThreadMessages = maxNumberOfThreadMessages;
+		_mbMessageLocalService = mbMessageLocalService;
+		_serviceContext = serviceContext;
+	}
+
+	public String getMessageBody(MBMessage mbMessage, String quoteMark) {
+		if (!_htmlFormat) {
+			return _getQuotedMessage(true, mbMessage.getBody(), quoteMark);
+		}
+
+		if (!mbMessage.isFormatBBCode()) {
+			return mbMessage.getBody();
+		}
+
+		try {
+			String messageBody = BBCodeTranslatorUtil.getHTML(
+				mbMessage.getBody());
+
+			HttpServletRequest httpServletRequest =
+				_serviceContext.getRequest();
+
+			if (httpServletRequest == null) {
+				return messageBody;
+			}
+
+			ThemeDisplay themeDisplay =
+				(ThemeDisplay)httpServletRequest.getAttribute(
+					WebKeys.THEME_DISPLAY);
+
+			return MBUtil.replaceMessageBodyPaths(themeDisplay, messageBody);
+		}
+		catch (Exception exception) {
+			_log.error(
+				StringBundler.concat(
+					"Unable to parse mbMessage ", mbMessage.getMessageId(),
+					": ", exception.getMessage()));
+		}
+
+		return mbMessage.getBody();
+	}
+
+	public String renderMessageParent(MBMessage parentMessage) {
+		if (!FeatureFlagManagerUtil.isEnabled("LPS-182020") ||
+			(_maxNumberOfThreadMessages == 0) ||
+			(parentMessage.getParentMessageId() ==
+				MBMessageConstants.DEFAULT_PARENT_MESSAGE_ID)) {
+
+			return StringPool.BLANK;
+		}
+
+		List<MBMessage> mbMessages = new ArrayList<>();
+
+		int numberOfMessages = _maxNumberOfThreadMessages;
+
+		while ((numberOfMessages > 0) && (parentMessage != null) &&
+			   (parentMessage.getParentMessageId() !=
+				   MBMessageConstants.DEFAULT_PARENT_MESSAGE_ID)) {
+
+			mbMessages.add(0, parentMessage);
+
+			parentMessage = _mbMessageLocalService.fetchMBMessage(
+				parentMessage.getParentMessageId());
+			numberOfMessages--;
+		}
+
+		_numMessagesByParentMessageId.put(
+			parentMessage.getMessageId(), mbMessages.size());
+
+		if (ListUtil.isEmpty(mbMessages)) {
+			return StringPool.BLANK;
+		}
+
+		StringBundler sb = new StringBundler(mbMessages.size());
+
+		sb.append(_getMarkupElement(MarkupElement.START_MESSAGE_THREAD));
+
+		int endElementCount = 0;
+
+		for (int i = 0; i < mbMessages.size(); i++) {
+			MBMessage mbMessage = mbMessages.get(i);
+
+			sb.append(
+				StringBundler.concat(
+					_getMarkupElement(MarkupElement.START_MESSAGE),
+					_getMarkupElement(MarkupElement.START_USER_MESSAGE),
+					_getUserName(mbMessage, _getQuote(i)),
+					_getMarkupElement(MarkupElement.END),
+					_getMarkupElement(MarkupElement.START_MESSAGE_BODY),
+					getMessageBody(mbMessage, _getQuote(i)),
+					_getMarkupElement(MarkupElement.END),
+					_getMarkupElement(MarkupElement.END_MESSAGE)));
+
+			endElementCount++;
+		}
```

Comment:
If you replace the `ArrayList` with a `LinkedList`, change this loop so that the code doesn't require random access by using a foreach loop and keeping track of `i` manually:
```
int i = 0;
for (MBMessage mbMessage : mbMessages) {
   ...
   i++;
}
```

Reply:
Also, remember to use names consistently. You have variables with both "mbMessage" and "message"; as this class is in an MB app bundle, either of those is fine. Pick one style and use it consistently.